crates/xhr/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+            XmlHttpRequestUpload {
+                upload: self.xhr.upload().unwrap_throw(),
+            }
+        }
+
+        /// The `XMLHttpRequest` method `open()` initializes a newly-created request, or
+        /// re-initializes an existing one.
+        ///
+        /// [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open)
+        pub fn open(&self, method: &http::Method, url: &str) {
+            self.xhr.open(&method.to_string(), url).unwrap_throw()
+        }
+
+        /// Abort the request.
+        ///
+        /// This will fire an `abort` event on the `XmlHttpRequest`.
+        pub fn abort(&self) {
+            self.xhr.abort().unwrap_throw()
+        }
+
+        /// Perform the request.
+        ///
+        /// Thise returns immediately - attach events handlers to be notified of the
+        /// progress and outcome of the request.
+        pub fn send<B: XhrBody>(&self, body: SendBody<B>) {
+            match body {
+                SendBody::None => self.send_without_body(),
+                SendBody::Body(body) => self.send_with_body(body),
+            }
+        }
+
+        /// `send()` without a body.
+        pub fn send_without_body(&self) {
+            self.xhr
+                .send()
+                .expect_throw("Error sending request. Did you forget to call `open`?")
+        }
+
+        /// Actually send the request. In order to know the outcome, you have to attach
+        /// `load`, `abort` and `error` event listeners.
+        pub fn send_with_body<B: XhrBody>(&self, body: B) {
+            body.send(&self.xhr)
+                .expect_throw("Error sending request. Did you forget to call `open`?")
+        }
+
+        /// Set a header on the request.
+        pub fn set_request_header(&self, header: &str, value: &str) {
+            self.xhr
+                .set_request_header(header, value)
+                .expect_throw("XHR header can be set")
+        }
+
+        /// Set the response timeout.
+        pub fn set_timeout(&self, timeout: std::time::Duration) {
+            self.xhr.set_timeout(timeout.as_millis() as u32)
+        }
+
+        /// Get a single header on the response.
+        pub fn get_response_header(&self, header: &str) -> Option<String> {
+            self.xhr.get_response_header(header).unwrap_throw()
+        }
+
+        /// Get all the response headers.
+        pub fn get_all_response_headers(&self) -> std::collections::HashMap<String, String> {
+            self.xhr
+                .get_all_response_headers()
+                .expect_throw("XMLHttpRequest.getAllResponseHeaders() error")
+                .split("\r\n")
+                .map(|line| {
+                    let mut elems = line.split(": ");
+                    (
+                        elems.next().unwrap_throw().to_string(),
+                        elems.next().unwrap_throw().to_string(),
+                    )
+                })
+                .collect()
+        }
```

Comment:
Would it make sense for this to be an [`http::HeaderMap`](https://docs.rs/http/0.1.17/http/header/struct.HeaderMap.html) instead?

Reply:
yep, this needs to be case insensitive, I realized writing tests for it